### ISSUE TITLE
Should 0.17 -> 0.18 upgrading be done automatically or only with a command-line argument?

### ISSUE THREAD

**Issue by avh4** [avh4/elm-format#258]
(Or should some things be done automatically, and other things only done once/manually?)
### Automatic

If you use the Elm-0.18 version of elm-format, and/or give the `--elm-version=0.18` command-line argument, then the upgrade transformations will be performed.
### Manual

Even if you use the Elm-0.18 version of elm-format, and/or give the `--elm-version=0.18` command-line argument, you will still have to use a special `--upgrade` command-line argument to do the upgrade transformations.  (You would generally only do so once.)

**Comment by zkessin** [comment]
I think manually, my gut says that there will folks out there who may wish to stay with 0.17 for some amount of time so having a `--upgrade=0.18` flag would probably be better

**Comment by avh4** [comment]
Ah, with "automatic", it would still only apply if you either 1) downloaded the Elm 0.18 version of elm-format, or 2) give the `--elm-version=0.18` command line argument.

**Comment by zkessin** [comment]
oh in that case I think it is fine.

**Comment by eeue56** [comment]
The problem with automatic:

- User starts project in 0.17
- User sets up elm-format in their editor etc
- User upgrades elm and elm-format to 0.18
- User still uses backticks
- Elm-format removes backticks and replaces them
- User continues to use backticks because they don't know it's invalid elm


This means that elm-format is doing metaprogramming transformations on my code. I expect elm-format to only do formatting changes, and not syntax changes. Applying such changes automatically breaks the " principle of least astonishment". An upgrade tool should be different in concept to a linter/formatting, even if they're built through the same tool.

**Comment by eeue56** [comment]
Summary of other approaches:

tl;dr - only rust automatically saves changes to files. Very few tools provide syntax changes beyond formatting. Those that do, all require flags to specify writing it.

## gofmt

https://golang.org/cmd/gofmt/

- By default, does not save changes to the files to format. 
- Use `-w` flag to write changes to the files.
- `-s` to "simplify" code according to some rewrite rules

## rubocop

http://rubocop.readthedocs.io/en/latest/basic_usage/

- By default, does not save changes to the file on lint. Also explains in English the problem, but does not provide the alternative.
- use `-a` to save format fixes 
- Auto fixing/saving does not work perfectly

## 2to3, 3to2

https://docs.python.org/2/library/2to3.html

- By default, does not save changes to the file 
- use `-w` to upgrade code
- pass rules to upgrade certain parts like import 

## Flake8

https://flake8.pycqa.org/en/latest/user/invocation.html

- By default, does not save changes to the file
- select certain rules to enable/disable by choice
- Provides no way to automatically fix problems (pep8 does)

## Swift migrator

https://swift.org/migration-guide/

- When opening a new project in the newer version of xcode, it gives you the option to migrate the project to the latest version of swift
- Otherwise, you may choose to convert some code manually via the GUI

## rust-fmt

https://github.com/rust-lang-nursery/rustfmt

- By default, overwrites and saves to file
- Does not help with upgrading syntax, only formatting

## cljfmt

https://github.com/weavejester/cljfmt

- By default, only lints
- pass the `fix` argument to make it auto-fix and save

## jshint

http://jshint.com/docs/

- By default, only lints and has no ability to modify anything

**Comment by klazuka** [comment]
I would prefer that upgrade is a manual step. The name of the tool, elm-format, does not imply that it does anything beyond formatting Elm code. If you are a user of elm-format and are not following this conversation, you're probably going to be pretty surprised when it gets done behind your back.

I've now converted my project twice to Elm 0.18: the first time by hand and the second time with the assistance of elm-format. The latter experience was *much better* (thank you), although the business of fixing syntax errors file-by-file was a pain (and eventually I ended up using `find` and `xargs` to invoke elm-format across all .elm files in the project).

By making the migration process explicit, it also frees the tool up to do more things:
- migrate *all* files in the project
- warn the user before continuing when they have deps that are not 0.18 compatible
- fixup the `elm-package.json` file
- blow-away `elm-stuff` artifacts
- etc.

**Comment by avh4** [comment]
Processing all files can already be done with `elm-format .` (will recursively find all *.elm files (excluding the elm-stuff folder))

I don't think there will be a tool for automatically upgrading deps for 0.18 / fixing elm-package.json / clearing elm-stuff.  I don't think I'll have time to think about how that could be done by elm-format, and I don't know of anyone else working on such a tool.

elm-format has been doing automatic upgrading of 0.16 code to 0.17, and there were no complaints about that.  However, the 0.18 upgrade is different because there are significant library changes in addition to the language syntax changes.  I think the syntax changes to belong in elm-format; but the library changes do make sense to be a one-time processing step.  (And also, in the future, other major package releases might want to provide automated upgrades without relying on an elm-format release.)

For the moment, I'm going to make a list of additional package upgrades that we'd like to automate and see if they're easily implementable; and once that's done we can decide about changing how the upgrade is triggered.

**Comment by klazuka** [comment]
> Processing all files can already be done with elm-format .

Neat. I missed it because it's not documented when you envoke `elm-format --help` (but it is documented in the README). I'll prepare a PR to fix that.

> I don't think there will be a tool for automatically upgrading deps for 0.18

Oh, I didn't mean for 0.18. My point was that the explicit upgrade flag sets up a path towards a more feature-rich upgrade tool in the future.

> I'm going to make a list of additional package upgrades that we'd like to automate...

For me, the biggest remaining item that makes the 0.18 upgrade tedious is dealing with the removal of `(:=)`. The problem is that it (1) occurs frequently and (2) cannot be fixed with a simple find-and-replace because the right-hand-side arg to the binary operator may need additional parenthesis to achieve the right precedence after switching to `field`.

**Comment by avh4** [comment]
> For me, the biggest remaining item that makes the 0.18 upgrade tedious is dealing with the removal of (:=). The problem is that it (1) occurs frequently and (2) cannot be fixed with a simple find-and-replace because the right-hand-side arg to the binary operator may need additional parenthesis to achieve the right precedence after switching to field

I think this is a good idea.  -> https://github.com/avh4/elm-format/issues/268

**Comment by avh4** [comment]
Regarding the original question of this thread, I still don't entirely agree, but I went ahead and implemented it anyway (See https://github.com/avh4/elm-format/issues/267).  The `--upgrade` flag will be required for the 0.18 upgrade.
